AI rebate decision

stackpoint opened the pull request:
Rebates the AI for being over the leader limit. AI is currently bugged to consider only generals when rolling to leader limit.

The num_of_X conditionals for the generals/admirals/conquistadors/explorers are bugged. The count starts at 1 instead of 0.

It seems like the game uses num_of_generals to determine if can roll for another general and doesn't consider admirals/conquistadors/explorers.

The num_of_generals conditional also counts free leaders from ruler/heir.

I couldn't think of a better way for the AI to get their points back since there is no monthly monarch point gain function that I could find.

stackpoint committed:
AI rebate decision

Rebates the AI for being over the leader limit. AI is currently bugged to consider only generals when rolling to leader limit.

Grociu commented:
If I'm reading it correctly:
- the AI is getting a decision to click once a year if they have 2 admiral, 2 explorers, 2 conquistadors.
- NumGen counts how many total leaders the AI has substracting the ruler and heir
- subtracts 4 from that number to simulate a free leader from each category (To account for the conditional bug. From your explanation, I'm assuming the value for example num_of_admirals is 1 if they have no admirals)
- then the FreeLeaderPool is compared to adjusted NumGen
- AI gets MIL points if that number is negative or 0

Just to visualize the issue you're trying to solve: 
- AI has 5 leader limit, and 4 generals and an admiral
- AI wants another admiral, sees that it has a leader slot free (counts only generals) and rolls an admiral.
- AI now pays 12 MIL a year for upkeep of an extra leader, but that leader was created only because decision programming here is buggy.
- AI clicks decision, gets 12 MIL rebate, decision will pop up in 365 days if condition (2 admirals) is valid.

Edge case:
At final count in the end, if I'm reading this right, you give the 12 MIL at value of 0. From my understanding this means that if they AI is at the leader cap, they get the points back? Why?

stackpoint commented:
> the AI is getting a decision to click once a year if they have 2 admirals, 2 explorers, 2 conquistadors.

The num_of_X conditionals for the generals/admirals/conquistadors/explorers are bugged. The count starts at 1 instead of 0 so they only get the option to click this if they have 1 admiral/explorer/conquistador.


>  At final count in the end, if I'm reading this right, you give the 12 MIL at value of 0. From my understanding this means that if they AI is at the leader cap, they get the points back? Why?

Remember that for paradox conditionals: X = 0 checks X greater than or equal to 0 and NOT = { X = 0 } checks X less than 0. So for this specific edge case it doesn't use 0.

The reason I used the negative comparison and NOT was because the free leader pool is a float that gets rounded down to the nearest integer. So if I have 13.99 free leader pool, I can only have 13 leaders.


numGen | numFLP | delta | NOT delta
-- | -- | -- | --
9 | 8.00 | -1.00 | TRUE
9 | 8.50 | -0.50 | TRUE
9 | 8.99 | -0.01 | TRUE
9 | 9.00 | 0.00 | FALSE
9 | 9.50 | 0.50 | FALSE
9 | 9.99 | 0.99 | FALSE

stackpoint committed:
Update AIExplorerRebate.txt

It turns out the trigger itself is not bugged but the export to variable function for the conditionals are.

stackpoint commented:
> The num_of_X conditionals for the generals/admirals/conquistadors/explorers are bugged.

It turns out the trigger itself is not bugged but the export to variable function for the conditionals are.

SpiritBaker reviewed (commented):
In my experience this bug only occurs while AI at war, and major powers almost constantly at war in late stages of the game. It slightly complicates things, because the issue is dynamic.

stackpoint commented:
I've also noticed that this does definitely happen more often when the AI are at war but does still occurs when they are at peace too. Use can use [this script](https://forumcontent.paradoxplaza.com/public/569465/run.txt) using the run console command to find all the countries over their leader limit in your vanilla games too.

Since this also happens in vanilla, I made a bug report on the paradox forums here: https://forum.paradoxplaza.com/forum/threads/ai-exceeding-leader-limit.1393832/
